### PR TITLE
Add TensorFlow Wav2Vec2 for sequence classification

### PR DESCRIPTION
**Pull request by nandwalritik**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #21778 
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@sanchit-gandhi 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sgugger** [comment]
Kindly ping @sanchit-gandhi and adding @Rocketknight1 for the TensorFlow side.

**Comment by Rocketknight1** [comment]
Hi @nandwalritik, and sorry for the extremely long delay in catching this! Ordinarily one of the TF maintainers reviews TF pull requests, but this one slipped through the cracks somehow. If you want to file TF PRs in future, you can directly ping me or @gante to make sure that we don't miss it.

This PR actually looks almost perfect, but there are a couple of TF-specific details that are causing some tests to fail. I'll mark them in a code review in just a sec, but they shouldn't take too long to fix. Thanks again for submitting this!

**Review by Rocketknight1 — COMMENTED** [review]
This looks good! A few tweaks in the `__init__` should fix most of the issues.

The only other thing missing is a `serving` and `serving_output` method. These are properties that are unique to TF models, and indicate the input and output signatures to enable model compilation and exporting. You can look at other models in the library to get a sense for how they work, but if you can't figure it out let me know and I'll add it for you!

File: src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py [hunk 1]
```diff
@@ -1552,3 +1592,69 @@ def serving_output(self, output: TFCausalLMOutput) -> TFCausalLMOutput:
...
         return TFCausalLMOutput(logits=output.logits, hidden_states=hidden_states, attentions=attentions)
+
+
+class TFWav2Vec2ForSequenceClassification(TFWav2Vec2PreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+        self.wav2vec2 = TFWav2Vec2Model(config)
+        num_layers = config.num_hidden_layers + 1
+        if config.use_weighted_layer_sum:
+            self.layer_weights = tf.Variable(tf.ones(shape=(num_layers,)) / num_layers)
```
[the remark this answers is not in the document]
Comment (nandwalritik) [reply src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py]: I added `layer_weights` initialization in `build` method, but when I try to print output for this
```
for layer in model_tf.layers:
    for weight in layer.weights:
        print(weight.name,weight.shape,tf.reduce_sum(tf.abs(weight)))
```
I am unable to get layer_weights, except this all the names and shapes of weights are matching.

**Comment by nandwalritik** [comment]
> 

for `serving` and `serving_output` methods I added changes, but now sure they are correct or not.

**Comment by Rocketknight1** [comment]
Hi @nandwalritik, I'm seeing the issue when you move it to `build()` - the problem is the weight name, as it usually is in our TensorFlow ports! TF isn't very consistent about the name scope used for weights, and it can differ depending on when the weight is created in the `init`, the `build` or lazily in the `call()`, which makes it tricky because we use the names to match weights between PT and TF models.

I'll see if I can push a solution to your repo, hang on.

**Comment by nandwalritik** [comment]
Ok

**Comment by Rocketknight1** [comment]
Try:

```
with tf.name_scope(self._name_scope()):
    self.layer_weights = self.add_weight(
        shape=(self.num_layers,), initializer="ones", trainable=True, name="layer_weights"
    )
```
in the `__init__`, not the `build()`. I know that contradicts what I said earlier, but it turns out to be a bit different for a base model class than a sublayer.

I also see a couple of other errors - you can see them by clicking the `Details` beside `tests_tf` in the checklist at the bottom of this PR. If you can't figure out what's causing them, ping me over the weekend or on Monday and I'll try to debug them!

**Comment by nandwalritik** [comment]
> Try:
> 
> ```
> with tf.name_scope(self._name_scope()):
>     self.layer_weights = self.add_weight(
>         shape=(self.num_layers,), initializer="ones", trainable=True, name="layer_weights"
>     )
> ```
> 
> in the `__init__`, not the `build()`. I know that contradicts what I said earlier, but it turns out to be a bit different for a base model class than a sublayer.
> 
> I also see a couple of other errors - you can see them by clicking the `Details` beside `tests_tf` in the checklist at the bottom of this PR. If you can't figure out what's causing them, ping me over the weekend or on Monday and I'll try to debug them!

Ok, so after adding this change, the weights are getting loaded without any warning or error, but the output of pytorch and tensorflow model doesn't have `rtol` of `1e-5`.
Although I checked shape and absolute sum of tensors of both the models they are almost equal
```
PT model 
1,292,768 -> 29877.8750


1,292,256 -> 29711.7109

pooled_output
1,256 -> 38.7491



TF model

hidden_state
1,292,768 -> 29877.879

1,292,256 -> 29711.715

pooled_output
1,256 -> 38.811996
```
What should i try next to satisfy rtol criteria.

**Comment by Rocketknight1** [comment]
Hm, those are some fairly large discrepancies! The debugging process we recommend when something like that happens is:

- Make a test environment and load the PT and TF models with the same weights
- Try to isolate the earliest point where the model outputs diverge. You can use options like `output_hidden_states` to get the model to return all hidden states, not just the final ones.
- Once you find the first point of divergence, try to see if you can dig into the layer where the divergence happened. You can place breakpoints, or extract sublayers and try passing test inputs into them.
- Eventually you will find the single specific place where the divergence creeps in - now you can check what the cause is. Make sure the weights for that operation really do match between the two framework, and make sure both frameworks are doing the same thing at that point.

As always, if you can't figure it out, let me know! This kind of work can be quite gruelling, but we really appreciate the work you're doing on the model port.

**Comment by nandwalritik** [comment]
Hi @Rocketknight1 I added test cases and fixed the feed forward part, but the CI is failing due to `flax`, I think this might not be related to my changes. Please review the PR and let me know if any more changes are required.

**Comment by Rocketknight1** [comment]
Yep, those flax issues are unrelated, just ignore them. I'll review everything today, but the CI looks good!

File: tests/models/wav2vec2/test_modeling_tf_wav2vec2.py [hunk 2]
```diff
@@ -244,6 +250,33 @@ def check_ctc_loss(self, config, input_values, *args):
...
+            attention_mask[i, input_lengths[i] :] = 0
+
+        # disable dropout and batch normalization
+        tf.keras.backend.set_learning_phase(0)
+
+        masked_loss = model(input_values, attention_mask=attention_mask, labels=labels).loss.numpy().item()
+        unmasked_loss = model(input_values, labels=labels).loss.numpy().item()
+
+        # enable dropout and batch normalization
+        tf.keras.backend.set_learning_phase(1)
```
Comment (Rocketknight1) [inline tests/models/wav2vec2/test_modeling_tf_wav2vec2.py:274]: `set_learning_phase` isn't wrong, but in the rest of the library we usually just pass a `training` bool to the models. Can you modify the layer to take that argument and pass it to the other layers, and then replace `set_learning_phase` here with `training=False`?

File: src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py [hunk 3]
```diff
@@ -1212,6 +1212,46 @@ def serving(self, inputs):
...
+        )  # these two operations makes sure that all values before the output lengths idxs are attended to
+        ## check device
+        attention_mask = tf.tensor_scatter_nd_update(
+            attention_mask,
+            indices=tf.stack([tf.range(batch_size), output_lengths - 1], axis=1),
+            updates=tf.ones([batch_size], dtype=attention_mask.dtype),
+        )
+        attention_mask = tf.reverse(attention_mask, axis=[-1])
+        attention_mask = tf.cumsum(attention_mask, axis=-1)
+        attention_mask = tf.reverse(attention_mask, axis=[-1])
```
Comment (Rocketknight1) [inline src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:1251]: `cumsum` has a `reverse` argument that might let you save a couple of lines here

**Review by Rocketknight1 — APPROVED** [review]
This looks really good! I left a couple of minor comments, but this is basically ready to go at this point. The inference tests give me very high confidence that this matches the behaviour of the PT/FLAX model up to numerical error. Thanks for all the effort you put in with this PR - it's really appreciated, and I think people will get a lot of use out of the result!

**Review by sanchit-gandhi — APPROVED** [review]
Very nice PR @nandwalritik - looks good from an audio perspective. Just wanted to confirm that the slow tests pass? I'm pretty confident we have equality between the TF and PT code based on what I've seen, but would love to hear what @Rocketknight1 says here too!

**Edit:** he beat me to it!

File: src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py (hunk 3 above)
Comment (nandwalritik) [reply src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:1251]: but that does it along axis=-1 I want to perform reverse operation on axis=[-1], I tried with that it was giving incorrect output.

Comment (Rocketknight1) [reply src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:1251]: Ah, okay! In that case, keep it this way.

**Comment by nandwalritik** [comment]
@sanchit-gandhi @Rocketknight1 let me know if any more changes are required or else can you guys get this pr merged.

**Comment by Rocketknight1** [comment]
Just looked over the last few changes - I'm happy to merge it at this point. Thanks again for putting in the work on this!
